Introducing the new Dataset Module for cleanlab 2.0

cgnorthcutt opened the pull request:
# Dataset module

This module introduces the new API for interacting and working with the entire datasets, with suggestions for overlapping classes that you may want to merge, low quality classes you may want to remove, overall dataset health score, and a fully health summary report (generate for your dataset in a single line of code).

* Code should probably be tested a bit more thoroughly before merge.
* Merge or rebase this commit. **Do not squash and merge.**

# need to add tests!
* please add tests as you review.

# Updates made to module during this PR:
* Merged the two label quality and label noise methods.
* report conditional matrix scores (instead of joint) for class methods
* All methods that previously returned multiple objects now work with and return pandas dataframes.
* Added support for string classes
* Added a health_summary function that prints everything about your dataset health in one line of code.
* completely rewrote the docstrings.
* all methods take in all input types.
* all methods have been renamed.
* Added testing for four real-world datasets.

# Future support

* in the future we should add support for all the functions from the raw data directly + a model (like `CleanLearning`), but that's for a later release.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+    3. labels and pred_probs
+    but it is not necessary to provide a combination, e.g. the labels and pred_probs and joint.
+
+    For parameter info, see the docstring of `count.estimate_joint`
```
The docstring for `count.estimate_joint` does not seem to explain some of these args like `joint`, `asymmetric`. Furthermore the docstring of `count.estimate_joint` does not really explain its parameters either: 
https://docs.cleanlab.ai/master/cleanlab/count.html#cleanlab.count.estimate_joint
It just says:  See `cleanlab.count.calibrate_confident_joint` docstring.

jwmueller replied:
Even if `count.estimate_joint` did have good explanations of its parameters, I would prefer to explain the parameters separately here.  The reason is that `count.estimate_joint` seems like a more advanced function that requires understanding concepts, whereas `find_similar_classes` is supposed to be a super accessible function easily used by anybody.

cgnorthcutt replied:
added params to both

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def find_similar_classes(
+    joint=None,
+    asymmetric=True,
+    *,
+    labels=None,
+    pred_probs=None,
```
I find this argument order a bit peculiar.
Consider having the arg order be like this match the other parts of package:
```
labels=None,
pred_probs=None,
*,
joint=None,
confident_joint=None,
asymmetric=True,
```

In particular, I'd think about it like this: 100% of cleanlab users will know what `labels` and `pred_probs` are. But only say 50% of users will know what `joint` and `confident_joint` are.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+from count import estimate_joint, compute_confident_joint
+
+
+def find_similar_classes(
```
Should we replace "similar" -> "commonly_confused" like we had in our cleanlab-studio report?

One reason is similar_classes just sounds like something I could get from a confusion matrix from model's predictions.

cgnorthcutt replied:
I would think "confused" sounds more like it comes from a "confusion" matrix 🤣

jwmueller replied:
Ya that is indeed annoying as well... 

Also sorry I didn't even meant confusion matrix in my criticism of "similar" but rather all that's needed is model's pred_probs to find similar classes (can just measure correlation in pred_probs).

Alternatives names for this function could include:
`find_conflated_classes`
`find_swapped_classes`
`label_error_types`
`mislabeling_rates`

cgnorthcutt replied:
how about `find_classes_to_merge` ?

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+    The method uses the joint distribution of noisy and true labels to compute ontological issues
+    via the approach published in (Northcutt, Jiang, Chuang (2021) JAIR).
+    paper: https://arxiv.org/abs/1911.00068
+
```
I think we want to clarify how this method differs from the traditional confusion matrix of a model's predictions.

Eg. while confusion matrix in traditional classification measures how similar a model thinks two classes are, this function measures how often the annotators confuse two classes.

cgnorthcutt replied:
done.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+
+    Returns
+    -------
+        A tuple of 2 ordered lists: (list_of_pairs_of_classes, likelihood_they_should_be_merged)"""
```
likelihood_they_should_be_merged is not well-defined.
I think it may be overly bold to claim "should_be_merged".

Two classes could be fundamentally confusing to annotators while still not mergeable (eg. "benign tumor" vs "malignant tumor" in medical images).

jwmueller replied:
What is advantage of this return type vs say:
List[tuple] where each tuple = (class1, class2, confusion_score)

Or pd.DataFrame outlined in above comment.

jwmueller replied:
the confusion_score (what you're calling likelihood_they_should_be_merged) should have some intuitive definition of what it's value represents right?

cgnorthcutt replied:
Agree. Any recommendations? It's based on the joint probability.

jwmueller replied:
Hm... as a user the information that I'd be most interested in is:
234 examples labeled as class A probably should actually be labeled as class B instead.

So I'd print that at least. 
If scores  = joint probability, you could explain it as:
"score(A,B) = The proportion of examples in the dataset that are labeled as class A but probably should actually be labeled as class B instead."

cgnorthcutt replied:
I implemented conditional probability scores for the class level functionality and added thorough explanations in all Returns sections

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+      If `asymmetric==False`, the pair (class1, class2) will onyl be returned once and order is
+      arbitrary (internally this is just summing score(class1, class2) + score(class2, class1)).
+
+    Returns
```
Consider adding `verbose` arg, and also having this function print out the info in a pretty format.

jwmueller replied:
Eg. I'd personally have this function organize this info in a pd.DataFrame something like this:
```
class1  class2  confusion_score
A          B           0.95
```
That can then easily be printed in nice format (and manipulated in various ways, which the current returned structure is harder to manipulate).

In future could consider plotting it as a heatmap, just like folks normally do with confusion matrix.

jwmueller replied:
For the top few most-confused class-pairs, it'd also be nice to print more info like:

cleanlab found 124 examples with given label A that should actually be labeled as class B

cgnorthcutt replied:
My gut also was to return a dataframe. I didn't because right now Cleanlab doesn't depend on pandas. Can you think of any downsides to introducing pandas as a new dependency? I kept it to just numpy and small fast packages to keep Cleanlab lean as long as possible. Thoughts @anishathalye and @jwmueller ?

jwmueller replied:
I think it'll eventually have to support pandas in order to avoid having all sorts of arrays/lists all over tracking different data issues. Will probably add pandas dependency in 2.1 once multiple data issues are being flagged in a single run. And most data scientists will anyway already be using pandas (I've never personally had issue getting pandas to play nice with another package).  

For v2.0, I don't have strong preference, but it does seem like the return types are being somewhat sacrificed by not having pandas. And if we anyway want to add pandas dependency  in v2.1, then at that point would not want to be stuck with this suboptimal return type.

jwmueller replied:
Another option is to lazy import pandas inside each function with an informative error msg if it fails, so pandas is only strictly needed for user to utilize these `dataset` functions.

cgnorthcutt replied:
made pandas a dependency for cleanlab. method now returns pd.DataFrame

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+    asymmetric : bool (default: True)
+      If `asymmetric==True`, includes both pairs (class1, class2) and (class2, class1). Use this for
+      finding "is a" relationships where for example "class1 is a class2".
+      If `asymmetric==False`, the pair (class1, class2) will onyl be returned once and order is
```
typo: onyl

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+      If `asymmetric==True`, includes both pairs (class1, class2) and (class2, class1). Use this for
+      finding "is a" relationships where for example "class1 is a class2".
+      If `asymmetric==False`, the pair (class1, class2) will onyl be returned once and order is
+      arbitrary (internally this is just summing score(class1, class2) + score(class2, class1)).
```
score(class1, class2) needs to be defined somewhere. Ideally it should have intuitive definition (like "proportion of examples with given label A that should likely be actually labeled as B instead")

cgnorthcutt replied:
i added very thorough docstrings for the Returns section that details all scores in depth

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+        # Find indices of similar classes
+        idx = np.array(list(zip(*np.unravel_index(np.argsort(matrix.ravel()), matrix.shape)))[::-1])
+        # Bring max score between 0.1 and 1.0, otherwise score scales inversely with num_classes ^ 2
+        scaling_factor = 10 ** np.floor(-np.log10(np.max(matrix)))
```
Ideally the score has intuitive definition, which I'm not sure this rescaled one does?

jwmueller replied:
why not just have the scores simply be this?

P(true_label = i | given_label = j)

That seems very interpretable to me

cgnorthcutt replied:
I would think p(given|true) is more intuitive since typically you condition on the latent truth? How did you make that choice? Also why do you like the conditional over the joint? The joint is more expressive albeit the numbers are tiny.

I agree the scaling factor is somewhat arbitrary, it just makes it a good experience since the numbers will always be something readable, but definitely plan to improve it and happy to immediately. Or simply remove it.

jwmueller replied:
I thought conditioning on given is more natural because that is what user sees in their data.
So eg. I know that if I see example with given label A, it is untrustworthy.

I'm less sure what I could do with the information that many examples truly should be labeled B but are labeled something else (beyond realizing that my data is under-representing the occurrence of class B).

Re conditional vs joint, I'll leave it up to you. I think both are useful depending on the user's goal: joint is more useful if I want to maximally improve my dataset by fixing A->B mislabels that are overall common. conditional is more useful if I want to compare which given label is most/least trustworthy.

cgnorthcutt replied:
I like the idea of implementing the noise matrix version of the scoring (because the number will be normalized and bigger), but the problem is that when `symmetric==True`, it lose its meaning since this i computed by summing the P(x,y) + P(y, x) which doen't make sene if its p(x|y) + P(y|x). Added this as something to think about for 2.1 and moving on with using the joint.

Removed the scaling entirely though, so it will return small joint probability score.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+    return class_indices, scores
+
+
+def get_classes_ranked_by_label_quality(joint, descending=True):
```
could shorten to: `rank_classes_by_label_quality`

jwmueller replied:
why not take in the similar args like `find_similar_classes()` does?

That function seems far more flexible. In particular, I think 100% of users will know about `labels` and `pred_probs` but many many not know what is `joint` (so I prefer to make this optional if user supplies labels & pred_probs)

cgnorthcutt replied:
good idea. both edits made.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+        return idx[:cutoff], scores[:cutoff]
+
+
+def get_classes_ranked_by_label_noise(joint, descending=True, rank_by="given_label"):
```
All my comments below for `get_classes_ranked_by_label_quality` apply to this method as well

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+
+def get_classes_ranked_by_label_quality(joint, descending=True):
+    """Returns the ordered class indices and associated scores quantifying the quality of the labels
+     in each class (larger number implies higher quality labels) for the joint of a dataset.
```
Clarify what is different about this method vs `get_classes_ranked_by_label_noise`

cgnorthcutt replied:
ended up completely refactoring and removing this method entirely and wrapping its functionality into a single method `rank_classes_by_label_quality`

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+    return class_indices, scores
+
+
+def dataset_label_quality_score(joint):
```
could consider alternative name `overall_label_quality_score` or `global_label_quality_score`

jwmueller replied:
again consider allowing many of the same args as `find_similar_classes`

cgnorthcutt replied:
Changed to `overall_label_health_score` and added args.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def dataset_label_quality_score(joint):
+    """Returns a single metric for the quality of the labels in a dataset. The score is between
```
"the labels" -> "all labels"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+def dataset_label_quality_score(joint):
+    """Returns a single metric for the quality of the labels in a dataset. The score is between
+    0 and 1. A higher score implies higher quality labels, with 1 implying labels that have no
+    issues."""
```
data scientists will appreciate if you provide an intuitive definition of what the score means

cgnorthcutt replied:
added

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+    """Returns a single metric for the quality of the labels in a dataset. The score is between
+    0 and 1. A higher score implies higher quality labels, with 1 implying labels that have no
+    issues."""
+
```
consider verbose arg and printing some interesting info like:
"Overall 3.6% of the labels in the dataset are estimated to have issues"

cgnorthcutt replied:
added

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+"""Dataset Module
+Supports dataset-level and class-level automated quality, including finding which classes should
+be merged with other classes in your dataset, which classes should be removed,
+and which classes have the best label accuracy, and which have the worst labels.
```
"the best label accuracy, and which have the worst labels"
->
"tend to be annotated most/least correctly overall"

cgnorthcutt replied:
changed

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+
+"""Dataset Module
```
consider demonstrating these functions at end of one of the quickstart tutorials, maybe `text.ipynb`

cgnorthcutt replied:
I won't have time for this on my end, but leave this up for grabs! xD

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,150 @@
+Supports dataset-level and class-level automated quality, including finding which classes should
+be merged with other classes in your dataset, which classes should be removed,
+and which classes have the best label accuracy, and which have the worst labels.
+"""
```
It could be nice if all the functions in this file support optional arg:
```class_names = None```
where user can provide mapping between string names for the classes and int representations.
Eg. `class_names` supported inputs could be dict and list:
```
{0: 'classA', 1: 'classB',...}
```
and 
```
['classA', 'classB']
```

Then if `class_names is not None`, all print statements / outputs are presented in terms of class names instead of ints.

This can also be for future PR.

cgnorthcutt replied:
it actually even simpler! just needs to be a list with the class names in order (because we have guarantees on the format of `labels`).

jwmueller replied:
I was suggesting either list or dict input formats could be supported for this function (i.e. both types work).  When I use cleanlab, I personally prefer to track the mapping between class-names and ints as a dict in my own code before running cleanlab commands, because dict is much more interpretable when I print it.

cgnorthcutt replied:
Went ahead and implemented this for 2.0 🎉

cgnorthcutt commented:
IMPORTANT - this merge should include the commit history (do not squash, either merge directly or rebase and merge).

cgnorthcutt commented:
## Module completely rewritten.

* Merged the two label quality and label noise methods.
* report conditional matrix scores (instead of joint) for class methods
* All methods that previously returned multiple objects now work with and return pandas dataframes.
* Added support for string classes
* Added a health_summary function that prints everything about your dataset health in one line of code.
* completely rewrote the docstrings.
* all methods take in all input types.
* all methods have been renamed.
@anishathalye @jwmueller

cgnorthcutt commented:
Example output of all methods (below)
![image](https://user-images.githubusercontent.com/27030270/162625526-55e04014-a804-45d0-b1d0-cc672fb53a44.png)

Example output of health summary in one line of code for CIFAR-100:
![image](https://user-images.githubusercontent.com/27030270/162625582-cc84d513-e35c-4230-a23d-11b75ce74c60.png)

cgnorthcutt commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+        num_issues = (num_examples * (1 - joint_trace)).round().astype(int)
+        print(
+            f" * Overall, about {1 - joint_trace:.0%} ({num_issues:,} of the {num_examples:,}) "
+            f"labels in your dataset have issues.\n"
```
Changed to `labels in your dataset **may potentially have** issues`

jwmueller replied:
[Nit] "may potentially" can be shortened to "potentially"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+
+"""Dataset Module
```
[Nit] None of the other modules descriptions start like this: 
Dataset Module

Cf: https://docs.cleanlab.ai/master/cleanlab/rank.html

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+
+
+"""Dataset Module
+Supports dataset-level and class-level automated quality, including finding which classes should
```
could be clarified: "Supports dataset-level and class-level automated quality, including ..."

Example alternative: "Provides dataset-level and class-level overviews of the label issues in your data, including ..."

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+        confident_joint=confident_joint,
+        multi_label=multi_label,
+    )
+    print("Overall Class Quality and Noise across your dataset (below)")
```
maybe instead of "(below)" could say something more descriptive like:
"(sorted below with lowest quality labels near the top)"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    )
+    print("Overall Class Quality and Noise across your dataset (below)")
+    print("-" * 60, "\n", flush=True)
+    print(df_class_label_quality)
```
should we just print `df.head()` instead? The topmost ones seem most important to review

jwmueller replied:
this applies to all dataframe prints throughout this module

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+        + "\n",
+    )
+
+    df_class_label_quality = rank_classes_by_label_quality(
```
Some of the DF column names could be more clear if using terms from our Studio report-template instead.

Eg:

"Overall Label Issues" -> "Num. Label Issue"

jwmueller replied:
Could add informative column from our Studio report template:
"% of Examples with Label Issue" 
which is supposed to = 100 * (num-examples with this given label and flagged as having issue) / (num-examples with this given label)

jwmueller replied:
Also could add informative column from our Studio report template:
"Number of examples given this label"
which helps user trade-off between prioritizing classes that are very noisy but rare overall vs classes that are somewhat noisy but very common overall.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    )
+    if class_names is not None:
+        df.insert(loc=0, column="Class Name", value=class_names)
+    return df.sort_values(by="Overall Label Noise Score", ascending=False)
```
[Nit] future development will be less bug-prone if you define global variable:

OVERALL_LABEL_NOISE_SCORE = "Overall Label Noise Score"
and have all code refer to OVERALL_LABEL_NOISE_SCORE instead throughout.  This would ideally be done for all strings referenced in multiple places of the code.

anishathalye replied:
Great idea, will do later

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+            The index of the class in 0, 1, ..., num_classes - 1.
+        * Overall Label Issues
+            count(given_label = k, true_label != k)
+            Estimated number of label issues in the class (usually the most accurate method).
```
unclear what this means: "(usually the most accurate method)"

jwmueller replied:
Consider rename: "Overall Label Issues" -> "Number of Label Issues" or "Num. Label Issues"

jwmueller replied:
"number of label issues in the class" could be clarified more:

"number of label issues among the examples whose given label is this class"

jwmueller replied:
[Nit] I believe the way you have formatted docstring may not render well in the docs. May be better to format it like this:

* Overall Label Issues
            Estimated number of label issues in the class (usually the most accurate method), i.e.:  count(given_label = k, true_label != k)
            
Alternatively if you want count(...) to be on separate line, might need to insert additional line-break.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+        * Overall Label Issues
+            count(given_label = k, true_label != k)
+            Estimated number of label issues in the class (usually the most accurate method).
+        * Inverse Label Issues
```
Consider rename of "Inverse Label Issues"  to something more descriptive like: "Number of Missed Instances"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+            Estimated number of label issues in the class (usually the most accurate method).
+        * Inverse Label Issues
+            count(given_label != k, true_label = k)
+            Estimated number of times this class occurs in other classes in the dataset.
```
"Estimated number of times this class occurs in other classes in the dataset." could be more clear as something like:
"Estimated number of examples for which this class is more appropriate label than their given label."

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+        * Inverse Label Issues
+            count(given_label != k, true_label = k)
+            Estimated number of times this class occurs in other classes in the dataset.
+        * Overall Label Noise Score
+            prob(true_label != k | given_label = k)
+            Estimated proportion of label issues in the class.
+        * Inverse Label Noise Score
+            prob(given_label != k | true_label = k)
+            Estimated proportion of times this class label occurs in other classes in the dataset.
+        * Overall Label Quality Score
+            p(given_label = k | true_label = k)
+            How accurate the labels are in each class as a conditional prob p(given=k | true=k).
```
Do we really need all 3 of these?  I found it a little overwhelming to see so many scores whose definition I didn't know as a user.

jwmueller replied:
At least "Overall Label Quality Score" and "Overall Label Noise Score" convey the same information AFAICT. Would just keep "Overall Label Quality Score" to stick with same convention as rest of the package.

jwmueller replied:
"Inverse Label Noise Score" is derivable from "Inverse Label Issues" and the other columns. So want to ask ourself, is precomputing this score for the user worth potentially overwhelming them a little?  I felt a little  overwhelmed by seeing so many scores.

jwmueller replied:
Finally anytime the nebulous term "Score" is used, we probably want to stick with convention throughout that lower score -> more serious problem that warrants user's attention.  It is major QOL improvement for user if they don't have to remember which scores are higher = better vs lower = better.

jwmueller replied:
Label Quality Score can be very clearly defined as follows:

How accurate the labels tend to be overall for a given class, i.e. the estimated probability that the true label should be `k` averaged across all examples whose given label = `k`.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+        * Overall Label Quality Score
+            p(given_label = k | true_label = k)
+            How accurate the labels are in each class as a conditional prob p(given=k | true=k).
+        By default, the DataFrame is ordered by "Overall Label Noise Score" descending.
```
Shouldn't the whole cleanlab package stick with the convention:
 lower scores -> more serious problem that warrants attention
 like we did for our current (per-example) label-quality scores?
I think we need to stick with one convention throughout, otherwise user has to remember what higher/lower means for every score and we'll soon have tons of scores.

anishathalye replied:
Removed "Score", so now it's just "Overall Label Noise"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+        print(
+            f" * Overall, about {1 - joint_trace:.0%} ({num_issues:,} of the {num_examples:,}) "
+            f"labels in your dataset have issues.\n"
+            f" ** The overall label health score for this dataset is: {joint_trace:.2f}."
```
can be clearer if you add: 
XXX (scores range between 0-1, higher is better)

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    This method provides two scores in the pandas Data Frame that is returned:
+    * "Num Overlapping Examples" - The number of examples where the two classes overlap
+    * "Joint Probability" - "Num Overlapping Examples" / total number of examples in the dataset.
+    interchangeable and returns a dataframe with the classe and the joint probability score
```
typos in text and formatting issue

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+
+    This method provides two scores in the pandas Data Frame that is returned:
+    * "Num Overlapping Examples" - The number of examples where the two classes overlap
+    * "Joint Probability" - "Num Overlapping Examples" / total number of examples in the dataset.
```
I like the score "joint probability" because it's very interpretable. However one issue is this violates the convention of our package that lower scores => more serious problem. 

If we want to be consistent with this convention (which I advise given we plan to have tons of data-quality scores in the package), then alternate option is to call this "Overlap Score" and in the docstring define it as 1 - joint probability, i.e. = 1 - (Number of overlapping examples) / (Total number of examples in the dataset)

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    -------
+    pd.DataFrame
+        A Pandas DataFrame with columns "Class Index A", "Class Index B",
+        "Num Overlapping Examples", "Joint Probability" and a description of each below.
```
Add: "Each row corresponds to a pair of classes."

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    pd.DataFrame
+        A Pandas DataFrame with cols "Class Index", "Overall Label Issues", "Inverse Label Issues",
+        "Overall Label Issues", "Inverse Label Noise Score", "Overall Label Quality Score" and a
+        description of each below. Noise scores are between 0 and 1, where 0 implies no label issues
```
Add after first sentence: "Each row corresponds to one class label."

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    multi_label=False,
+    verbose=True,
+):
+    """Returns a single score/metric between 0 and 1 for the quality of all labels in a dataset.
```
[Nit] "quality" -> "overall quality"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    verbose=True,
+):
+    """Returns a single score/metric between 0 and 1 for the quality of all labels in a dataset.
+    Intuitively, the score is the average label consistency across all classes in the
```
"correctness of the given labels" instead of "label consistency"?

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    confident_joint=None,
+    multi_label=False,
+):
+    """Prints a healthy summary of your datasets including results for powerful statistics like:
```
"healthy summary of your datasets including results for powerful statistics like"
->
"summary of your dataset's overall label health, including:"

jwmueller replied:
[Nit] There are a couple occurrences of "datasets" in this file. Would replace them with "dataset"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    """Prints a healthy summary of your datasets including results for powerful statistics like:
+    * the classes with the most and least label issues
+    * classes that overlap and could potentially be merged
+    * overall data label quality health score statistics for your dataset
```
[Nit] could delete final bullet since it's implied by "overall health summary"

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+
+    Parameters
+    ----------
+    For parameter info, see the docstring of `dataset.find_overlapping_classes`
```
I'd imagine `health_summary()` is the method we envision most users invoking from this module (and just getting results from find_overlapping_classes from the internal call inside health_summary).

So consider moving the docstrings for parameters to this method, and then having the Docstring for find_overlapping_classes say: See docstring of `health_score` instead.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    print("-" * 60, "\n", flush=True)
+    print(df_class_label_quality)
+
+    df_class_label_quality = find_overlapping_classes(
```
[Nit] confusing variable name choice, would name: `df_overlapping_classes` instead, especially if we want to have option to return both of these.

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,416 @@
+    print()
+    health_score = overall_label_health_score(labels=labels, pred_probs=pred_probs, verbose=True)
+    print("\nGenerated with <3 from Cleanlab.")
+    return health_score
```
`health_summary()` seems to already compute the results of every other function in this module, but they're being tossed. Any way to return these to user as well?

This speaks to the need for future `Dataset` class to store intermediate results. 

For now, this method could have boolean flag argument: `return_value :  List[str]` with default value: `["health_score", "class_label_quality", "class_overlap"]` (and another option for this list could be `"joint"`). Then 
based on `return_value`, this method returns a dict whose keys = the values in `return_value` and values = the corresponding scores / dataframes / arrays.

jwmueller reviewed (changes requested):
New output format looks great!  I mostly suggest some minor API refactors, and major changes in naming conventions to improve usability/clarity.

JohnsonKuan commented:
> Example output of all methods for Google Quickdraw dataset (below): ![Screen Shot 2022-04-10 at 7 31 09 AM](https://user-images.githubusercontent.com/27030270/162625732-bff68a81-7f5e-4be5-8b52-3ecafae39939.png)
> 
> Example output of health summary in one line of code for CIFAR-100: ![image](https://user-images.githubusercontent.com/27030270/162625582-cc84d513-e35c-4230-a23d-11b75ce74c60.png)

Output looks awesome.

It may be helpful to explain in the docstring why the estimated joint prob of `(A, B)` is not the same as `(B, A)` like in your quickdraw example where joint prob of `(birthday cake, cake)` differed from `(cake, birthday cake)`. I know why they differ (after initial read of the confident learning paper) but I'm guessing most new users will probably not know.

jwmueller commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,426 @@
+import io
+import numpy as np
+from cleanlab import dataset
+
```
May want test for edge-case with only 2 classes as well?  I could envision potential bugs / formatting issues in that edge-case

cgnorthcutt replied:
yup included in current tests (imdb is 2 classes)

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,411 @@
+    print("\nGenerated with <3 from Cleanlab.")
+    return {
+        "health_score": health_score,
+        "joint": estimate_joint,
```
BUG:  This "joint" key was a function for me in my test run:

<function estimate_joint at 0x1236dc4c0>

Does not seem right

cgnorthcutt replied:
i think you fixed this. thanks

jwmueller commented on cleanlab/dataset.py:
```diff
@@ -0,0 +1,411 @@
+        df.insert(
+            loc=1, column="Class Name B", value=df["Class Index B"].apply(lambda x: class_names[x])
+        )
+    return df.sort_values(by="Joint Probability", ascending=False)
```
May want to be careful about the dataframe's index (ie. reset it).
Currently looks arbitrary in my run:

```
Class Overlap. In some cases, you may want to merge classes in the top rows (below)
-----------------------------------------------------------------------------------

  Class Name A Class Name B  Class Index A  Class Index B  Num Overlapping Examples  Joint Probability
7            C            D              2              3                         4               0.04
9            D            E              3              4                         3               0.03
3            A            E              0              4                         1               0.01
6            B            E              1              4                         1               0.01
0            A            B              0              1                         0               0.00
1            A            C              0              2                         0               0.00
2            A            D              0              3                         0               0.00
4            B            C              1              2                         0               0.00
5            B            D              1              3                         0               0.00
8            C            E              2              4                         0               0.00
```

jwmueller replied:
The index for classes_by_label_quality returned DataFrame seems to make sense though AFAICT, it's simply the class index.

cgnorthcutt commented:
Addressed all comments and added thorough testing on real datasets with 2, 10, and more classes. Verified that methods return correct values in some cases.
